src/cargo/util/config/mod.rs

```diff
@@ -545,9 +560,30 @@ impl Config {
                 return true;
             }
         }
+        self.check_environment_key_case_mismatch(key);
+
         false
     }
 
+    fn check_environment_key_case_mismatch(&self, key: &ConfigKey) {
+        if cfg!(windows) {
+            // In the case of windows the check for case mismatch in keys can be skipped
+            // as windows already converts its environment keys into the desired format.
+            return;
+        }
+
+        match self.upper_case_env.get(key.as_env_key()) {
+            Some(env_key) => {
+                let _ = self.shell().warn(format!(
+                    "Environment variables require uppercase letters, \
```

Comment:
Maybe the language can be improved here: "Environment variables are expected to use uppercase letters and underscores, the variable `{}` will be ignored and have no effect".